### PR TITLE
EndSession support

### LINKED ISSUE #912: explicit closing of sessions
With the newly introduced multi-session support, there are some usecases where we want to use a session for a short time and then free up the cache slot for further uses.

* when we want to reprompt for a passphrase (e.g. in Trezor Suite empty wallet confirmation), we need to allocate a new session. Every reprompt potentially evicts one of the valid passphrased sessions.
* when enumerating connected devices, we should (although don't _need_ to, technically speaking) initialize a session to get Features. repeated enumerations would also needlessly evict sessions

We propose a new message, `EndSession`, that cancels the current session, erases its ID and removes it from the cache.
Subsequent calls without `Initialize` would return a new failure code, `InvalidSession`. For ease of implementation, we could allow calls that do not touch session data (e.g., device settings), and return the failure only for calls that use the seed.

### PR DESCRIPTION
**Pull request by matejcik**
this PR introduces an implementation of the `EndSession` message and fixes #912 

the new behavior is as follows:

- until first `Initialize` message is received, any attempt to use cached data is rejected with a `ProcessError("Invalid session")`
- after `Initialize`, a session is allocated
- calling `EndSession` destroys data of that session, and resets to the "invalid session" state. Subsequent uses of cache will again raise a `ProcessError`

This allows us to reuse session cache slots: a session ended with `EndSession` is cleared away, so we can have 9 "long-lived" sessions and reuse the 10th slots for things like requesting passphrase again. Suite should start using it ASAP.

`trezorctl` is also updated to make use of this: every command ends the session after itself, unless the session was resumed by `trezorctl -s <session_id>`

---

open questions:
- should we introduce a new failure code, `InvalidSession`, as suggested in the original bug? ISTM there is low possibility of breakage, as previously the call would end in a FirmwareError anyway.
- should we enforce session validity in _all_ calls, or just the ones that touch the cache? It would be cleaner, but ISTM there is no actual need to do it.

### CODE REVIEW

**Comment by matejcik** [comment]
there is a behavior change on legacy: previously, calling e.g. `GetAddress` without a valid session would silently allocate one. Now it is necessary to call `Initialize` explicitly.

**Comment by matejcik** [comment]
Randomness has changed again in the TT PIN tests. I originally thought this is because we now call `client.clear_session()` unconditionally, but the affected tests are those where `client.clear_session()` would be called either way. It's probably nothing but I'll need to look into it more closely anyway.

**Comment by tsusanka** [comment]
The only breaking change for hosts is that we currently require `Initialize` beforehand. Right? And EndSession is not "mandatory"

**Comment by tsusanka** [comment]
> should we introduce a new failure code, InvalidSession, as suggested in the original bug? ISTM there is low possibility of breakage, as previously the call would end in a FirmwareError anyway.

Please, what is the original bug?

**Comment by tsusanka** [comment]
Btw is this going to work with HWI? Because they are trying not to use the session id or something like that aren't they?

**Review by tsusanka — APPROVED** [review]
Approve for _Core_.

**Comment by matejcik** [comment]
> The only breaking change for hosts is that we currently require `Initialize` beforehand and they should be using that already anyway. Right? And EndSession is not "mandatory".

It's only breaking on T1. On TT since 2.3.0 doing anything without `Initialize` would result in a FirmwareError anyway.

other than that, yes. well-behaved clients will see no changes, and it is not necessary to use `EndSession`

> Please, what is the original bug?

i meant the original issue, it's not a bug :) #912 

> Btw is this going to work with HWI? Because they had some problems with our sessions didn't they?

i believe so, but we should do some tests anyway

**Review by mmilata — COMMENTED** [review]
Not very familiar with the session handling but for the untrained eye the change is looking good :+1:

File: core/tests/test_storage.cache.py [hunk 1]
```diff
@@ -128,6 +161,15 @@ def call_Initialize(**kwargs):
         call_Initialize(session_id=session_id)
         self.assertEqual(cache.get(KEY), "hello")
 
+    def test_EndSession(self):
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/tests/test_storage.cache.py:164]: mock_storage is only required for `Initialize`, which calls `apps.base.get_features`, which looks into storage fields

**Review by prusnak — APPROVED** [review]

**Review by mmilata — COMMENTED** [review]
Btw maybe also consider documenting `EndSession` message in `docs/common/communication/sessions.md`?

**Pull request merged**